### PR TITLE
soroban-rpc: preflight: Exclude temporary expired entries from SnapshotSource

### PR DESCRIPTION
**Pull request by 2opremio**

### CODE REVIEW

**Review by leighmcculloch — APPROVED** [review]

**Review by leighmcculloch — APPROVED** [review]
Is this something we can include a test for?

**Comment by 2opremio** [comment]
Hold on, this is still broken

**Comment by 2opremio** [comment]
Alright, this should be it. 

> Is this something we can include a test for?

Probably, but it's 1 am here :) , feel free to add one

File: cmd/soroban-rpc/lib/preflight/src/ledger_storage.rs [hunk 1]
```diff
@@ -227,28 +245,27 @@ impl LedgerStorage {
...
-                Ok(le) => le,
-            };
-        if let Some(ref tracker) = self.restore_tracker {
-            _ = tracker.track_and_restore(self.current_ledger_sequence, key, &entry_and_expiration);
+        let result = <dyn SnapshotSource>::get(self, key);
+        if let Err(ref host_error) = result {
+            if host_error.error == HostError::from(Error::NotFound).error {
+                return Ok(false);
+            }
         }
```
Comment (2opremio) [inline cmd/soroban-rpc/lib/preflight/src/ledger_storage.rs:268]: I am not sure this is the best way (or even a correct way) to compare Host errors

@dmkozh / @leighmcculloch  can you take a look?

Comment (dmkozh) [reply cmd/soroban-rpc/lib/preflight/src/ledger_storage.rs:268]: Typically we verify host errors wit hhost_error.error.is_code/is_type, but for that you'll need to know the mapping from Error::NotFound.

File: cmd/soroban-rpc/lib/preflight/src/ledger_storage.rs [hunk 2]
```diff
@@ -227,28 +245,27 @@ impl LedgerStorage {
...
+            let mut entry_and_expiration = self
+                .get_including_persistent_expired(key)
+                .map_err(HostError::from)?;
             // If the entry expired, we modify the expiration to make it seem like it was restored
             entry_and_expiration.1 =
                 tracker.track_and_restore(self.current_ledger_sequence, key, &entry_and_expiration);
+            return Ok((entry_and_expiration.0.into(), entry_and_expiration.1));
         }
-        Ok((entry_and_expiration.0.into(), entry_and_expiration.1))
+        let entry_and_expiration =
```
Comment (dmkozh) [inline cmd/soroban-rpc/lib/preflight/src/ledger_storage.rs:257]: Just wanted to double-check: what is the error that occurs when 'auto-restoration' is not enabled and the persistent entry has expired? We need to make sure it's not `NotFound` as to distinguish from a simply non-existent entry.

File: cmd/soroban-rpc/lib/preflight/src/ledger_storage.rs (hunk 1 above)
Comment (2opremio) [reply cmd/soroban-rpc/lib/preflight/src/ledger_storage.rs:258]: So, is the code correct?

File: cmd/soroban-rpc/lib/preflight/src/ledger_storage.rs (hunk 2 above)
Comment (2opremio) [reply cmd/soroban-rpc/lib/preflight/src/ledger_storage.rs:247]: The error is also`NotFound`. (`Missing` for the host).

Yeah, we should probably use a different error code but:
1. This is for a separate PR.
2. I don't think the HostError has a suitable error to express it

File: cmd/soroban-rpc/lib/preflight/src/ledger_storage.rs (hunk 1 above)
Comment (2opremio) [reply cmd/soroban-rpc/lib/preflight/src/ledger_storage.rs:258]: I found a way to improve it. Thanks

File: cmd/soroban-rpc/lib/preflight/src/ledger_storage.rs (hunk 2 above)
Comment (2opremio) [reply cmd/soroban-rpc/lib/preflight/src/ledger_storage.rs:247]: Actually, I bit the bullet and included a separate error code for expired entries

Comment (dmkozh) [reply cmd/soroban-rpc/lib/preflight/src/ledger_storage.rs:251]: Thanks for addressing this as well.

**Review by dmkozh — APPROVED** [review]

**Review by leighmcculloch — APPROVED** [review]
I'm not an expert in this area of code, but the changes look like they make sense.

**Pull request merged**